File: ts-src/hooks/matcher.ts
import type { HookContext, ResolvedHookDefinition } from "./types.ts";

/**
 * Check if a hook's matchers match the given context.
 * No matchers = always matches. All matchers use AND logic.
 */
export function matchesHook(hook: ResolvedHookDefinition, context: HookContext): boolean {
	const { matchers } = hook;
	if (!matchers) return true;

	// tool matcher
	if (matchers.tool !== undefined) {
		if (!context.toolName) return false;
		if (Array.isArray(matchers.tool)) {
			if (!matchers.tool.includes(context.toolName)) return false;
		} else {
			if (matchers.tool !== context.toolName) return false;
		}
	}

	// match_path: glob against file_path extracted from toolArgs JSON
	if (matchers.match_path !== undefined) {
		if (!context.toolArgs) return false;
		let filePath: string | undefined;
		try {
			const parsed = JSON.parse(context.toolArgs);
			filePath = parsed.file_path;
		} catch {
			return false;
		}
		if (!filePath) return false;
		const glob = new Bun.Glob(matchers.match_path);
		if (!glob.match(filePath)) return false;
	}

	// match_args: glob against entire toolArgs string
	if (matchers.match_args !== undefined) {
		if (!context.toolArgs) return false;
		const glob = new Bun.Glob(matchers.match_args);
		if (!glob.match(context.toolArgs)) return false;
	}

	// match_input: glob against userInput
	if (matchers.match_input !== undefined) {
		if (!context.userInput) return false;
		const glob = new Bun.Glob(matchers.match_input);
		if (!glob.match(context.userInput)) return false;
	}

	return true;
}

File: src/hooks/matcher.rs
//! Hook matcher predicates.

use globset::Glob;
use serde_json::Value;

use super::types::{HookContext, ResolvedHookDefinition, ToolMatch};

fn glob_match(pattern: &str, target: &str) -> bool {
    match Glob::new(pattern) {
        Ok(g) => g.compile_matcher().is_match(target),
        Err(_) => false,
    }
}

pub fn matches_hook(hook: &ResolvedHookDefinition, ctx: &HookContext) -> bool {
    let matchers = match &hook.matchers {
        Some(m) => m,
        None => return true,
    };

    if let Some(tool) = &matchers.tool {
        let tool_name = match ctx.tool_name.as_deref() {
            Some(n) => n,
            None => return false,
        };
        match tool {
            ToolMatch::Single(s) => {
                if s != tool_name {
                    return false;
                }
            }
            ToolMatch::Many(list) => {
                if !list.iter().any(|s| s == tool_name) {
                    return false;
                }
            }
        }
    }

    if let Some(path_pat) = &matchers.match_path {
        let tool_args = match ctx.tool_args.as_deref() {
            Some(a) => a,
            None => return false,
        };
        let parsed: Result<Value, _> = serde_json::from_str(tool_args);
        let file_path = match parsed {
            Ok(v) => v
                .get("file_path")
                .and_then(|x| x.as_str())
                .map(String::from),
            Err(_) => return false,
        };
        let fp = match file_path {
            Some(p) => p,
            None => return false,
        };
        if !glob_match(path_pat, &fp) {
            return false;
        }
    }

    if let Some(args_pat) = &matchers.match_args {
        let tool_args = match ctx.tool_args.as_deref() {
            Some(a) => a,
            None => return false,
        };
        if !glob_match(args_pat, tool_args) {
            return false;
        }
    }

    if let Some(input_pat) = &matchers.match_input {
        let input = match ctx.user_input.as_deref() {
            Some(i) => i,
            None => return false,
        };
        if !glob_match(input_pat, input) {
            return false;
        }
    }

    true
}
